omegaconf/base.py

```diff
@@ -413,15 +416,71 @@ def _resolve_interpolation_from_parse_tree(
                 raise
             return None
 
-        assert resolved is not None
-        if isinstance(resolved, str):
-            # Result is a string: create a new StringNode for it.
-            return StringNode(
-                value=resolved,
-                key=key,
-                parent=parent,
-                is_optional=value._metadata.optional,
-            )
+        # If the output is not a Node already (e.g., because it is the output of a
+        # custom resolver), then we will need to wrap it within a Node.
+        must_wrap = not isinstance(resolved, Node)
+
+        if isinstance(value, AnyNode):
+            # If the node is untyped, we still need to validate against `None`
+            # in case the `optional` flag is set to `False`.
+            if not value._is_optional() and OmegaConf.is_none(resolved):
+                if throw_on_resolution_failure:
+                    self._format_and_raise(
+                        key=key,
+                        value=resolved,
+                        cause=InterpolationValidationError(
+                            "Non optional field cannot be assigned None"
+                        ),
+                    )
+                return None
+
+        elif isinstance(value, ValueNode):
+            # If the node is typed, validate (and possibly convert) the result.
+            res_value = _get_value(resolved)
+            try:
+                conv_value = value.validate_and_convert(res_value)
+            except ValidationError as e:
+                if throw_on_resolution_failure:
+                    self._format_and_raise(
+                        key=key,
+                        value=res_value,
+                        cause=e,
+                        type_override=InterpolationValidationError,
+                    )
+                return None
+
+            # If the converted value is of the same type, it means that no conversion
+            # was actually needed. As a result, we can keep the original `resolved`
+            # (and otherwise, the converted value must be wrapped into a new node).
+            if type(conv_value) != type(res_value):
```

Comment:
Deducing it from type of the result type seems fragile.
Maybe validate_and_convert should return a second boolean flag indicating if it converted or not?
Alternatively, we can break it into two function:

```python
def is_valid_value(self, value) -> bool:
  ...

def convert_value(self, value) -> Any:
  ...

def validate_and_convert(self, value) -> Any:
  if self.is_valid_value(value):
    return value
  else:
    return self.convert_value(value)
```

This can be a sizeable change. If we do it I suggest we do it as a standalone diff independently of this diff. Once it's in - this diff can utilize the more fine-grained API instead of calling validate_and_convert().

Reply:
> Deducing it from type of the result type seems fragile.
> Maybe validate_and_convert should return a second boolean flag indicating if it converted or not?

Or, we impose the contract that it should return the same object if it is not converted -- back to my first version :)

> Alternatively, we can break it into two function

A potential concern with breaking into two functions is that it may do the same checks twice.

Let me know what you think makes most sense for this PR, and I'll go with it.

Reply:
I think having the same type checks twice during assignment is minor.

Breaking the function into two will also play better when we try to deprecate the automatic conversion on assignment of primitive types. (as opposed to automatic conversion on merge, which I think we want to keep).

Reply:
Ok but how should I proceed for this PR? I think there are 4 options:
1. Keep current check based on type
2. Go back to previous version where the check was based on the object returned being the same
3. Have `validate_and_convert()` return a boolean flag indicating whether a change was made
4. Split `validate_and_convert()` in two functions